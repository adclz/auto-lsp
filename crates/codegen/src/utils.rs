pub(crate) fn sanitize_string(string: &str) -> String {
    if let Some(v) = PUNCTUATION.get(string) {
        v.to_string()
    } else if let Some(v) = RUST_KEYWORDS.get(string) {
        v.to_string()
    } else {
        let mut result = String::new();
        for c in string.chars() {
            if c == '_' {
                result.push(c);
            } else if let Some(v) = PUNCTUATION.get(&c.to_string()) {
                result.push_str(v);
            } else {
                result.push(c);
            }
        }
        result.replace(" ", "_")
    }
}

pub(crate) fn sanitize_string_to_pascal(string: &str) -> String {
    if let Some(v) = PUNCTUATION.get(string) {
        v.to_string()
    } else if let Some(v) = RUST_KEYWORDS.get(string) {
        v.to_string()
    } else {
        let mut result = String::new();
        for c in string.chars() {
            if c == '_' {
                result.push(c);
            } else if let Some(v) = PUNCTUATION.get(&c.to_string()) {
                result.push_str(v);
            } else {
                result.push(c);
            }
        }
        result
            .replace(" ", "_")
            .split('_')
            .map(|word| {
                let mut chars = word.chars();
                match chars.next() {
                    Some(first) => format!(
                        "{}{}",
                        first.to_ascii_uppercase(),
                        chars.collect::<String>()
                    ),
                    None => String::new(),
                }
            })
            .collect()
    }
}

pub static PUNCTUATION: phf::Map<&'static str, &'static str> = phf::phf_map! {
    "{" => "LeftCurly",
    "}" => "RightCurly",
    "(" => "LeftParen",
    ")" => "RightParen",
    "[" => "LeftBracket",
    "]" => "RightBracket",
    "," => "Comma",
    ":" => "Colon",
    ";" => "Semicolon",
    "." => "Dot",
    "@" => "At",
    "!" => "Bang",
    "#" => "Hash",
    "$" => "Dollar",
    "%" => "Percent",
    "^" => "Caret",
    "&" => "Ampersand",
    "*" => "Star",
    "-" => "Minus",
    "_" => "Underscore",
    "+" => "Plus",
    "=" => "Equal",
    ">" => "Greater",
    "<" => "Less",
    "|" => "Pipe",
    "~" => "Tilde",
    "/" => "Slash",
    "\\" => "Backslash",
    "//" => "SlashSlash",
    "//=>" => "SlashSlashEqual",
    "//=" => "SlashSlashEqual",
    "/=" => "SlashEqual",
    "/>" => "SlashGreater",
    "/?" => "SlashQuestion",
    "/??" => "SlashNullish",
    "/*" => "SlashStar",
    "*/" => "StarSlash",
    "+++" => "PlusPlusPlus",
    "!!" => "BangBang",
    "!!=" => "BangBangEqual",
    "!!?" => "BangBangQuestion",
    "!!??" => "BangBangNullish",
    "!!???" => "BangBangNullishQuestion",
    "?" => "Question",
    "->" => "Arrow",
    "=>" => "FatArrow",
    "++" => "PlusPlus",
    "--" => "MinusMinus",
    "&&" => "And",
    "||" => "Or",
    "==" => "EqualEqual",
    "!=" => "NotEqual",
    ">=" => "GreaterEqual",
    "<=" => "LessEqual",
    "===" => "TripleEqual",
    "!==" => "NotTripleEqual",
    "<<" => "ShiftLeft",
    ">>" => "ShiftRight",
    ">>>" => "ShiftRightUnsigned",
    "+=" => "PlusEqual",
    "-=" => "MinusEqual",
    "*=" => "StarEqual",
    "%=" => "PercentEqual",
    "&=" => "AmpersandEqual",
    "|=" => "PipeEqual",
    "^=" => "CaretEqual",
    "&&=" => "AndEqual",
    "||=" => "OrEqual",
    "??=" => "NullishEqual",
    "??" => "Nullish",
    "???" => "NullishQuestion",
    "**" => "StarStar",
    "**=" => "StarStarEqual",
    "<>" => "LessGreater",
    "<=>" => "LessGreaterEqual",
};

pub static RUST_KEYWORDS: phf::Map<&'static str, &'static str> = phf::phf_map! {
    "abstract" => "Abstract",
    "as" => "As",
    "async" => "Async",
    "await" => "Await",
    "break" => "Break",
    "const" => "Const",
    "continue" => "Continue",
    "crate" => "Crate",
    "dyn" => "Dyn",
    "else" => "Else",
    "enum" => "Enum",
    "extern" => "Extern",
    "false" => "False",
    "final" => "Final",
    "fn" => "Fn",
    "for" => "For",
    "if" => "If",
    "impl" => "Impl",
    "in" => "In",
    "let" => "Let",
    "loop" => "Loop",
    "match" => "Match",
    "mod" => "Mod",
    "move" => "Move",
    "mut" => "Mut",
    "pub" => "Pub",
    "ref" => "Ref",
    "return" => "Return",
    "self" => "Self",
    "static" => "Static",
    "struct" => "Struct",
    "super" => "Super",
    "trait" => "Trait",
    "true" => "True",
    "type" => "Type",
    "unsafe" => "Unsafe",
    "use" => "Use",
    "where" => "Where",
    "while" => "While",
    "with" => "With",
    "yield" => "Yield",
    "None" => "_None",
    "Some" => "_Some",
    "Ok" => "_Ok",
    "Err" => "_Err",
    "Result" => "_Result",
    "Option" => "_Option",
    "Vec" => "_Vec",
};
