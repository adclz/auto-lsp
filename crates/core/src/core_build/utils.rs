/*
This file is part of auto-lsp.
Copyright (C) 2025 CLAUZEL Adrien

auto-lsp is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>
*/

pub(crate) fn intersecting_ranges(
    range1: &std::ops::Range<usize>,
    range2: &tree_sitter::Range,
) -> bool {
    range1.start <= range2.start_byte && range1.end >= range2.end_byte
}

pub(crate) fn tree_sitter_range_to_lsp_range(range: &tree_sitter::Range) -> lsp_types::Range {
    let start = range.start_point;
    let end = range.end_point;
    lsp_types::Range {
        start: lsp_types::Position {
            line: start.row as u32,
            character: start.column as u32,
        },
        end: lsp_types::Position {
            line: end.row as u32,
            character: end.column as u32,
        },
    }
}
