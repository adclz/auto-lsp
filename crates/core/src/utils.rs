#[macro_export]
macro_rules! dispatch_once {
    ($node:expr, [$($ty:ty => $method:ident($($param:expr),*)),*]) => {
        $(
            if let Some(n) = $node.downcast_ref::<$ty>() {
                return n.$method($($param),*);
            };
        )*
    };
}

#[macro_export]
macro_rules! dispatch {
    ($node:expr, [$($ty:ty => $method:ident($($param:expr),*)),*]) => {
        $(
            if let Some(n) = $node.downcast_ref::<$ty>() {
                n.$method($($param),*)?;
            };
        )*
    };
}