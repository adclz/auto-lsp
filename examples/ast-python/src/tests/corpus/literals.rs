/*
This file is part of auto-lsp.
Copyright (C) 2025 CLAUZEL Adrien

auto-lsp is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>
*/

use super::utils::Result;
use crate::snap;
use rstest::rstest;

#[rstest]
#[case("-1")]
#[case("0xDEAD")]
#[case("0XDEAD")]
#[case("1j")]
#[case("-1j")]
#[case("0o123")]
#[case("0O123")]
#[case("0b001")]
#[case("0B001")]
#[case("1_1")]
#[case("0B1_1")]
#[case("0O1_1")]
#[case("0L")]
fn integers(#[case] input: &str) -> Result {
    snap!(input)
}

#[rstest]
#[case("-.6_6")]
#[case("+.1_1")]
#[case("123.4123")]
#[case("123.123J")]
#[case("1_1.3_1")]
#[case("1_1.")]
#[case("1e+3_4j")]
#[case(".3e1_4")]
fn floats(#[case] input: &str) -> Result {
    snap!(input)
}

#[rstest]
#[case("1e322")]
#[case("1e-3")]
#[case("1e+3")]
#[case("1.8e10")]
#[case("1.e10")]
#[case("-1e10")]
fn scientific_notation_floats(#[case] input: &str) -> Result {
    snap!(input)
}

#[rstest]
#[case(r#""I'm ok""#)]
#[case(r#"'"ok"'"#)]
#[case(r#"UR'bye'"#)]
#[case(r#"b'sup'"#)]
#[case(r#"B"sup""#)]
#[case(r#"`1`"#)]
#[case(r#""\\"#)]
#[case(r#""/""#)]
#[case(
    r#""multiline \
string""#
)]
#[case(r#"b"\x12\u12\U12\x13\N{WINKING FACE}""#)]
#[case(r#""\xab\1\12\123\'\"\a\b\f\r\n\t\v\\""#)]
#[case(r#""\xgh\o123\p\q\c\d\e\u12\U1234""#)]
#[case(r#"f'\N{GREEK CAPITAL LETTER DELTA}'"#)]
fn strings(#[case] input: &str) -> Result {
    snap!(input)
}

#[rstest]
#[case(r#"'ab\x00cd'"#)]
#[case(r#""\n""#)]
#[case(r#"r'ab\x00cd'"#)]
#[case(r#"ur"\n""#)]
#[case(r#"fr"\{0}""#)]
#[case(r#"r"\\""#)]
#[case(
    r#"r'"a\
de\
fg"'"#
)]
fn raw_strings(#[case] input: &str) -> Result {
    snap!(input)
}

#[rstest]
#[case(
    r#"re.compile(r"(\n|\A)#include\s*['\"]"
           r"(?P<name>[\w\d./\\]+[.]src)['\"]")"#
)]
fn raw_strings_with_escaped_quotes(#[case] input: &str) -> Result {
    snap!(input)
}

#[rstest]
#[case(r#"f"a {b(f'c {e} d')} e""#)]
#[case(r#"f"""a"{b}c""""#)]
#[case(r#"f"""a""{b}c""""#)]
#[case(r#"f"a {{}} e""#)]
#[case(r#"f"a {b}}}""#)]
#[case(r#"f"a {{{b}""#)]
#[case(r#"f"a {{b}}""#)]
#[case(r#"f"a {{{b}}}""#)]
#[case(r#"f"{c,}""#)]
#[case(r#"f"{yield d}""#)]
#[case(r#"f"{*a,}""#)]
#[case(
    r#"def function():
    return f"""
{"string1" if True else
 "string2"}""""#
)]
#[case(
    r#"def test(self):
    self.assertEqual(f'''A complex trick: {
2  # two
}''', 'A complex trick: 2')"#
)]
fn format_strings(#[case] input: &str) -> Result {
    snap!(input)
}

#[rstest]
#[case(r#"f"a {b:2} {c:34.5}""#)]
#[case(r#"f"{b:{c.d}.{d.e}}""#)]
#[case(r#"f"{a:#06x}""#)]
#[case(r#"f"{a=}""#)]
#[case(r#"f"{a=:.2f}""#)]
#[case(r#"f"{value:{width + padding!r}.{precision}}""#)]
fn format_strings_with_format_specifiers(#[case] input: &str) -> Result {
    snap!(input)
}

#[rstest]
#[case(r#""\x12 \123 \u1234""#)]
fn unicode_escape_sequences(#[case] input: &str) -> Result {
    snap!(input)
}

#[rstest]
#[case("True")]
#[case("False")]
#[case("None")]
fn other_primitives(#[case] input: &str) -> Result {
    snap!(input)
}

#[rstest]
#[case(r#""one" "two" "three""#)]
fn concatenated_strings(#[case] input: &str) -> Result {
    snap!(input)
}

#[rstest]
#[case(
    r#""""
A double quote hello,
without double or single quotes.
""""#
)]
#[case(
    r#""""
A double quote "hello",
with double quotes.
""""#
)]
#[case(
    r#""""
A double quote 'hello',
with single quotes.
""""#
)]
#[case(
    r#"'''
A single quote hello,
without double or single quotes.
'''"#
)]
#[case(
    r#"'''
A single quote 'hello',
with single quotes.
'''"#
)]
#[case(
    r#"'''
A single quote "hello",
with double quotes.
'''"#
)]
#[case(
    r#""""
A double quote hello\n\
with an escaped newline\n\
and another escaped newline\n\
""""#
)]
fn multi_line_strings(#[case] input: &str) -> Result {
    snap!(input)
}

#[rstest]
#[case("[a, b, [c, d]]")]
#[case("[*()]")]
#[case("[*[]]")]
#[case("[*a]")]
#[case("[*a.b]")]
#[case("[*a[b].c]")]
#[case("[*a()]")]
fn lists(#[case] input: &str) -> Result {
    snap!(input)
}

#[rstest]
#[case("[a + b for (a, b) in items]")]
#[case("[a for b in c for a in b]")]
#[case("[(x,y) for x in [1,2,3] for y in [1,2,3] if True]")]
#[case("[a for a in lambda: True, lambda: False if a()]")]
fn list_comprehensions(#[case] input: &str) -> Result {
    snap!(input)
}

#[rstest]
#[case("{a: 1, b: 2}")]
#[case("{}")]
#[case("{**{}}")]
#[case("{**a}")]
#[case("{**a.b}")]
#[case("{**a[b].c}")]
#[case("{**a()}")]
fn dictionaries(#[case] input: &str) -> Result {
    snap!(input)
}

#[rstest]
#[case("{a: b for a, b in items}")]
#[case("{a: b for c in d for e in items}")]
fn dictionary_comprehensions(#[case] input: &str) -> Result {
    snap!(input)
}

#[rstest]
#[case("{a, b, c,}")]
#[case("{*{}}")]
fn sets(#[case] input: &str) -> Result {
    snap!(input)
}

#[rstest]
#[case("{a[b][c] for a, b, c in items}")]
#[case("{r for s in qs for n in ms}")]
fn set_comprehensions(#[case] input: &str) -> Result {
    snap!(input)
}

#[rstest]
#[case("()")]
#[case("(a, b)")]
#[case("(a, b, c,)")]
#[case("(print, exec)")]
fn simple_tuples(#[case] input: &str) -> Result {
    snap!(input)
}

#[rstest]
#[case("(a[b][c] for a, b, c in items)")]
#[case("dict((a, b) for a, b in d)")]
#[case("(a for b in c for d in e,)")]
#[case("(x for x in range(1, 10))")]
fn generator_expressions(#[case] input: &str) -> Result {
    snap!(input)
}
