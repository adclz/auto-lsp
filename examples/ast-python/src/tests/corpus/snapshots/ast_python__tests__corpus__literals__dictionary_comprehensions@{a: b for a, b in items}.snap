---
source: src/tests/corpus/literals.rs
expression: module
---
Module {
    children: [
        SimpleStatement(
            ExpressionStatement(
                ExpressionStatement {
                    children: [
                        Expression(
                            PrimaryExpression(
                                DictionaryComprehension(
                                    DictionaryComprehension {
                                        body: Pair {
                                            key: PrimaryExpression(
                                                Identifier(
                                                    Identifier {
                                                        [RANGE]
                                                    },
                                                ),
                                            ),
                                            value: PrimaryExpression(
                                                Identifier(
                                                    Identifier {
                                                        [RANGE]
                                                    },
                                                ),
                                            ),
                                            [RANGE]
                                        },
                                        children: [
                                            ForInClause(
                                                ForInClause {
                                                    left: PatternList(
                                                        PatternList {
                                                            children: [
                                                                Identifier(
                                                                    Identifier {
                                                                        [RANGE]
                                                                    },
                                                                ),
                                                                Identifier(
                                                                    Identifier {
                                                                        [RANGE]
                                                                    },
                                                                ),
                                                            ],
                                                            [RANGE]
                                                        },
                                                    ),
                                                    right: [
                                                        Expression(
                                                            PrimaryExpression(
                                                                Identifier(
                                                                    Identifier {
                                                                        [RANGE]
                                                                    },
                                                                ),
                                                            ),
                                                        ),
                                                    ],
                                                    [RANGE]
                                                },
                                            ),
                                        ],
                                        [RANGE]
                                    },
                                ),
                            ),
                        ),
                    ],
                    [RANGE]
                },
            ),
        ),
    ],
    [RANGE]
}
